### PR TITLE
`www-x-form-urlencoded` content support for status-code responses

### PR DESCRIPTION
**Pull request by ayeshLK**
Fixes [#2572](https://github.com/ballerina-platform/ballerina-standard-library/issues/2572)

## Examples
With this change, we could return `x-form-urlencoded` content from status-code responses:
```ballerina
type AcceptedResponse record {|
    *http:Accepted;
    string mediaType = mime:APPLICATION_FORM_URLENCODED;
    map<string> body;
|};

service /serviceEp on listenerEp {
    resource function get accepted() returns AcceptedResponse {
        // implement logic
        return {
            body: {
                "message": "Request is accepted by the server"
            }
        };
    }
}
```
## Checklist
- [X] Linked to an issue
- [ ] Updated the changelog
- [X] Added tests

### CODE REVIEW

File: ballerina-tests/tests/url_encoded_responses_from_service_test.bal [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+|};
+
+type ServerErrorResponse record {|
+    *http:InternalServerError;
+    string mediaType = mime:APPLICATION_FORM_URLENCODED;
+    map<string> body;
+|};
+
+final map<string> & readonly acceptedResponseBody = {
+    "message": "Request is accepted by the server"
```
Comment (shafreenAnfar) [inline ballerina-tests/tests/url_encoded_responses_from_service_test.bal:43]: Let's add more key/value pairs.

File: ballerina/http_connection.bal [hunk 2]
```diff
@@ -210,57 +213,92 @@ isolated function createStatusCodeResponse(StatusCodeResponse message, string? r
...
-            response.setETag(result);
+isolated function processAnydata(Response response, anydata payload, string? mediaType = ()) {
+    match mediaType {
+        mime:APPLICATION_FORM_URLENCODED => {
+            if payload is map<string> {
+                string result = retrieveUrlEncodedData(payload);
+                response.setTextPayload(result, mime:APPLICATION_FORM_URLENCODED);
+            } else {
+                panic error InitializingOutboundResponseError("unsupported content for application/x-www-form-urlencoded media type");
+            }
```
Comment (shafreenAnfar) [inline ballerina/http_connection.bal:269]: Lets remove the else part

File: ballerina-tests/tests/url_encoded_responses_from_service_test.bal (hunk 1 above)
Comment (ayeshLK) [reply ballerina-tests/tests/url_encoded_responses_from_service_test.bal]: Fixed with [b142eb145e9aeb850bab933f3da7dc99c68bd071](https://github.com/ballerina-platform/module-ballerina-http/pull/830/commits/b142eb145e9aeb850bab933f3da7dc99c68bd071)

File: ballerina/http_connection.bal (hunk 2 above)
Comment (ayeshLK) [reply ballerina/http_connection.bal:273]: Updated the logic as we discussed [a8394959d9c66356c9398959c74c263495ac19f7](https://github.com/ballerina-platform/module-ballerina-http/pull/830/commits/a8394959d9c66356c9398959c74c263495ac19f7)

**Review by shafreenAnfar — COMMENTED** [review]
LGTM!

**Review by chamil321 — DISMISSED** [review]

**Comment by TharmiganK** [comment]
Don't we have negative test cases for this?

File: ballerina/http_connection.bal [hunk 3]
```diff
@@ -248,19 +259,60 @@ isolated function setPayload(anydata payload, Response response, boolean setETag
...
-        panic error InitializingOutboundResponseError(errMsg + result.message(), result);
+        panic error InitializingOutboundResponseError(string `anydata to json conversion error: ${result.message()}`, result);
+    }
+    response.setJsonPayload(result);
+    if setETag {
+        response.setETag(result);
+    }
+}
+
+isolated function updateETag(Response response, anydata payload) {
```
Comment (TharmiganK) [inline ballerina/http_connection.bal:311]: Do we need this function? The etag setting improvement is reverted right.

**Review by TharmiganK — DISMISSED** [review]

File: ballerina/http_connection.bal (hunk 3 above)
Comment (ayeshLK) [reply ballerina/http_connection.bal]: Fixed with [c8a6e4a438948f85db3d1be5524b96f9202686be](https://github.com/ballerina-platform/module-ballerina-http/pull/830/commits/c8a6e4a438948f85db3d1be5524b96f9202686be)

**Review by TharmiganK — APPROVED** [review]

**Pull request merged**